src/garage/_environment.py

```diff
@@ -0,0 +1,201 @@
+    | observation_space     | The observation space specification             |
+    +-----------------------+-------------------------------------------------+
+    | spec                  | The environment specifications                  |
+    +-----------------------+-------------------------------------------------+
+    | render_modes          | The list of supported render modes              |
+    +-----------------------+-------------------------------------------------+
+
+    Example of a simple rollout loop:
+
+    .. code-block:: python
+
+        env = MyEnv()
+        policy = MyPolicy()
+        first_ts = env.reset()
+        env.visualize()  # visualization window opened
+
+        episode = [first_ts, ]
+        while not ts.last():
+            action = policy.get_action(episode[-1].next_observation)
+            episode.append(env.step(action))
+
+        env.close()  # visualization window closed
+
+    Make sure your environment is pickle-able:
+        Garage pickles the environment via the `cloudpickle` module
+        to save snapshots of the experiment. However, some environments may
+        contain attributes that are not pickle-able (e.g. a client-server
+        connection). In such cases, override `__setstate__()` and
+        `__getstate__()` to add your custom pickle logic.
+
+        You might want to refer to the EzPickle module:
+        https://github.com/openai/gym/blob/master/gym/utils/ezpickle.py
+        for a lightweight way of pickle and unpickle via constructor
+        arguments.
+
+    """
+
+    @property
+    @abc.abstractmethod
+    def action_space(self):
+        """akro.Space: The action space specification."""
+
+    @property
+    @abc.abstractmethod
+    def observation_space(self):
+        """akro.Space: The observation space specification."""
+
+    @property
+    @abc.abstractmethod
+    def spec(self):
+        """EnvSpec: The environment specification."""
+
+    @property
+    @abc.abstractmethod
+    def render_modes(self):
+        """list: A list of string representing the supported render modes.
+
+        See render() for a list of modes.
+        """
+
+    @abc.abstractmethod
+    def reset(self):
+        """Resets the environment and returns the first `TimeStep` of sequence.
+
+        Returns:
+            TimeStep: The first time step. Note that this `TimeStep`
+            will have attributes `observation`, `action`, `reward` as None
+            as they are undefined.
+
+        """
+
+    @abc.abstractmethod
+    def step(self, action):
+        """Steps the environment with the action and returns a `TimeStep`.
+
+        If the environment returned the last `TimeStep` of a sequence (either
+        of type TERMINAL or TIMEOUT) at the previous step, this call to
+        `step()` will start a new sequence and `action` will be ignored.
+
+        This method will also start a new sequence if called after the
```

Comment:
i don't agree with this stance. Instead, I think behavior if `reset()` has not been called should be undefined or lead to an exception.

Reply:
I agree. It should raise a `RuntimeError`.